### ISSUE TITLE
make `per_filter_config` specific configs work automatically for filters

### ISSUE THREAD

**Issue by rgs1** [envoyproxy/envoy#10164]
Currently filter writers need additional plumbing to allow their filter to do per vhost/route overrides. E.g.:

```
// Fetch the virtual host specific config.
const VHostSpecificFilterConfig* Filter::getSpecificConfig() const {
  if (!decoder_callbacks_->route() || !decoder_callbacks_->route()->routeEntry()) {
    return nullptr;
  }

  const auto* entry = decoder_callbacks_->route()->routeEntry();
  const auto* settings = entry->virtualHost().perFilterConfig("pinterest.my_filter");

  if (settings != nullptr) {
    auto* vhost_config = dynamic_cast<const VHostSpecificFilterConfig*>(settings);
    return vhost_config;
  }

  return nullptr;
}
```

and then you have to merge that, with the config object you got from the Filter Factory. It's not _that_ much logic, but it's a bit of hassle to a) do it for every filter and b) remember that you have to do it.

Thoughts on providing some plumbing to avoid repeating the above pattern (or thoughts on the general future of per vhost/route config overrides)?

cc: @mattklein123 @htuch @fishcakez

**Comment by dio** [comment]
Seems very useful.

**Comment by mattklein123** [comment]
Agreed, IMO this could be part of the filter callback API and/or the route API.

**Comment by yuval-k** [comment]
@rgs1 When I made some changes to ext_authz, I wrote a utility that does all of that for you:

https://github.com/envoyproxy/envoy/blob/5b7d3afc7b0f190d39e6a065662e646460183511/source/common/http/utility.h#L375

**Comment by rgs1** [comment]
@yuval-k ah nice, I did some of that manually for #10766. I'll follow-up with another PR to use those methods.
